### ISSUE TITLE
Calling is_read_closed() on a Ready from a TcpStream returns true even when its not closed when retrieving the Ready with stream.ready(Interest::READABLE)

### ISSUE THREAD

**Issue by asteinba** [tokio-rs/tokio#5098]
**Version**
tokio-bug-example v0.1.0
└── tokio v1.21.2
    └── tokio-macros v1.8.0 (proc-macro)

**Platform**
Linux xyz 5.17.0-1018-oem #19-Ubuntu SMP PREEMPT Wed Sep 21 09:54:58 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

**Description**
I'm not sure if I found a bug or I just did or understand something wrong. I found the following behavior:

Calling `is_read_closed()` on a `Ready` from a `TcpStream` returns true even when its not closed when calling `.ready()` only with Interest::READABLE and not Interest::READABLE | Interest::WRITEABLE (doing it also with Interest::WRITEABLE works as I expect it).

Minimal example that reproduces this behavior:

```rust
use std::io;
use tokio::io::{Interest};
use tokio::net::{TcpListener};

#[tokio::main]
async fn main()  {
    let listener = TcpListener::bind("127.0.0.1:1234").await.unwrap();
    loop {
        match listener.accept().await {
            Ok((mut stream, _addr)) => {
                println!("Client connected!");
                loop {
                    tokio::time::sleep(std::time::Duration::from_millis(1)).await;
                    let ready = stream.ready(Interest::READABLE).await.unwrap();
                    if ready.is_read_closed() {
                        println!("it says that read is closed");
                    }
                    if ready.is_readable() {
                        println!("but its still says it's readable, let's try to read");
                        let mut data = vec![0; 1024];
                        match stream.try_read(&mut data) {
                            Ok(n) => {
                                println!("And I can even read as you can see: {}", std::str::from_utf8(&data[..n]).unwrap());
                            }
                            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
                                continue;
                            }
                            Err(e) => {
                                panic!("Some error happened.");
                            }
                        }
                    }
                }
            }
            Err(e) => println!("couldn't get client: {:?}", e),
        }
    }
}
```
Cargo.toml
```toml
[package]
name = "tokio-bug-example"
version = "0.1.0"
edition = "2021"

[dependencies]
tokio = { version = "1.21.2", features = ["net", "macros", "rt-multi-thread", "time"] }
```
After starting the application I just connect to it using `telnet 127.0.0.1 1234` and start writing some lines like this:
```bash
~> telnet 127.0.0.1 1234
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
Hello from telnet
Another hello from telnet!
Last hello from telnet
```

I then get the following output from the application:
```
Client connected!
it says that read is closed
but its still says it's readable, let's try to read
And I can even read as you can see: Hello from telnet

but its still says it's readable, let's try to read
it says that read is closed
but its still says it's readable, let's try to read
And I can even read as you can see: Another hello from telnet!

but its still says it's readable, let's try to read
it says that read is closed
but its still says it's readable, let's try to read
And I can even read as you can see: Last hello from telnet

but its still says it's readable, let's try to read
```
So IMHO the "it says that read is closed" should never be visible. When I change the upper call `stream.ready(Interest::READABLE)` to `stream.ready(Interest::READABLE | Interest::WRITABLE)` I then get the following output:
```
Client connected!
but its still says it's readable, let's try to read
And I can even read as you can see: Again a hello from telnet

but its still says it's readable, let's try to read
but its still says it's readable, let's try to read
And I can even read as you can see: And one more!

but its still says it's readable, let's try to read
```
As you can see it never says its read closed. Thats what I would also expect calling .ready(Interest::READABLE).

**Comment by Darksonn** [comment]
It seems like the codepath eventually goes [here](https://github.com/tokio-rs/mio/blob/26941afc3e9807d226553d861fc5d9fc78062255/src/sys/unix/selector/epoll.rs#L214-L220).

Thoughts @Thomasdezeeuw?

**Comment by Thomasdezeeuw** [comment]
I would be interested in what events are returned by epoll/Mio. Could someone run with added logging here: https://github.com/tokio-rs/tokio/blob/992a16812258bae5192a997484ada046a88a6425/tokio/src/runtime/io/mod.rs#L194. Simply printing the events with `fmt::Debug` should be sufficient. I don't have time to do it myself today, maybe in the weekend.

**Comment by asteinba** [comment]
Sure (with Interest::READABLE):
```
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        0,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 0,
    },
}
Client connected!
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        1,
    ),
    readable: false,
    writable: true,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLOUT,
        u64: 1,
    },
}
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        1,
    ),
    readable: true,
    writable: true,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN|EPOLLOUT,
        u64: 1,
    },
}
it says that read is closed
but its still says it's readable, let's try to read
And I can even read as you can see: Hello

but its still says it's readable, let's try to read
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        1,
    ),
    readable: true,
    writable: true,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN|EPOLLOUT,
        u64: 1,
    },
}
it says that read is closed
but its still says it's readable, let's try to read
And I can even read as you can see: world

but its still says it's readable, let's try to read
```

And with Interest::READABLE | Interest::WRITABLE:
```
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        0,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 0,
    },
}
Client connected!
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        1,
    ),
    readable: false,
    writable: true,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLOUT,
        u64: 1,
    },
}
it says that read is closed
but its still says it's readable, let's try to read
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        2147483648,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 2147483648,
    },
}
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        2147483648,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 2147483648,
    },
}
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        2147483648,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 2147483648,
    },
}
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        1,
    ),
    readable: true,
    writable: true,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN|EPOLLOUT,
        u64: 1,
    },
}
but its still says it's readable, let's try to read
And I can even read as you can see: Hello

[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        2147483648,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 2147483648,
    },
}
but its still says it's readable, let's try to read
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        2147483648,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 2147483648,
    },
}
[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        1,
    ),
    readable: true,
    writable: true,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN|EPOLLOUT,
        u64: 1,
    },
}
but its still says it's readable, let's try to read
And I can even read as you can see: world

[/home/asteinba/dev/rust/tokio/tokio/src/runtime/io/mod.rs:176] event = Event {
    token: Token(
        2147483648,
    ),
    readable: true,
    writable: false,
    error: false,
    read_closed: false,
    write_closed: false,
    priority: false,
    aio: false,
    lio: false,
    details: epoll_event {
        events: EPOLLIN,
        u64: 2147483648,
    },
}
```

**Comment by Thomasdezeeuw** [comment]
Thanks @asteinba.

@Darksonn unless Tokio processes Mio events elsewhere and are weren't logged I don't see any `read_closed: true` logs. Is the conversion from `mio::Event` to `tokio::Ready` done properly?

**Comment by asteinba** [comment]
I tried to dig into it, but I really don't now anything about the tokio/mio internals so its just a super wild guess. As far as I can see the `Ready` that I receive from the `stream.ready(Interest::READABLE)` call is created here (but it seems that this code is not called when I call ready with Interest::READABLE | Interest::WRITEABLE):

https://github.com/tokio-rs/tokio/blob/ca9dd726b1c00736faa72df1885831545306a0a5/tokio/src/runtime/io/scheduled_io.rs#L512-L521

And the `from_interest` method always seems to set Ready:READ_CLOSED when its readable:

https://github.com/tokio-rs/tokio/blob/ca9dd726b1c00736faa72df1885831545306a0a5/tokio/src/io/ready.rs#L181-L195

Removing this line
https://github.com/tokio-rs/tokio/blob/ca9dd726b1c00736faa72df1885831545306a0a5/tokio/src/io/ready.rs#L186

causes `is_read_closed` to return false but then of course its always false. I have the feeling that either the `from_interest` method is wrong or that calling `from_interest` in the State::Done branch is wrong. But I don't understand the internals enough to judge about this.

The thing is when calling `stream.ready` with `Interest::READABLE | Interest::WRITEABLE` I don't see that the first code snippet (tokio/src/runtime/io/scheduled_io.rs#L512-L521) I linked in this comment is called, which is why `is_read_closed` returns a proper value. I couldn't find the code that is called instead.
